File: code/initZs.m
zs= struct('z',[]);

sz= size(z); n= sz(1); m= sz(2);
for i=1:9
    zs(i).z= z; 
end
xForw= [2:n,n]; xBack= [1,1:(n-1)];
yForw= [2:m,m]; yBack= [1,1:(m-1)];

zs(1).z= z(xForw, :   );
zs(2).z= z(xForw,yForw);
zs(3).z= z( :   ,yForw);
zs(4).z= z(xBack,yForw);
zs(5).z= z(xBack, :   );
zs(6).z= z(xBack,yBack);
zs(7).z= z( :   ,yBack);
zs(8).z= z(xForw,yBack);
zs(9).z= z(xForw, :   );

File: pycode/src/initZs.py
import numpy as np
from .schemas.zstruct import ZStruct
from .utils.list_utils import create_row_vector


def init_Zs(z):
    n, m = np.shape(z)
    xForw, xBack, yForw, yBack = __create_index_vectors(n, m)

    return [
        ZStruct(z[xForw]),
        ZStruct(z[xForw][:, yForw]),
        ZStruct(z[:, yForw]),
        ZStruct(z[xBack][:, yForw]),
        ZStruct(z[xBack]),
        ZStruct(z[xBack][:, yBack]),
        ZStruct(z[:, yBack]),
        ZStruct(z[xForw][:, yBack]),
        ZStruct(z[xForw])
    ]


def __create_index_vectors(n, m):
    # matlab [2:n, n]
    xForw = create_row_vector(1, n-1, post=n-1)
    # matlab [1,1:( n-1)]
    xBack = create_row_vector(0, n-2, pre=0)
    # matlab [2:m, m]
    yForw = create_row_vector(1, m-1, post=m-1)
    # matlab [1,1: (m-1)]
    yBack = create_row_vector(0, m-2, pre=0)
    return xForw, xBack, yForw, yBack
